cluster-autoscaler/proposals/plugable-provider-grpc.md

```diff
@@ -0,0 +1,127 @@
+# Plugable Cloud Provider over gRPC
+
+## Motivation
+
+CA is released as a bundle which includes a hardcoded list of supported cloud providers.
+Whenever users want to implement the logic of their own cloud provider, they need to fork the CA and add their own implementation.
```

Comment:
this is no longer true, a user could write a cluster-api provider without ever having to touch the autoscaler code.

Reply:
Agreed, but I'm not entirely comfortable having Cluster API as our default recommendation. There are certain problems that come with it's "cattle" approach to nodes differing from CA's design of operating on individual nodes (ex. best effort support for PDBs and annotations, no good intergration story for node upgrade).
Perhaps most importantly Cluster API seems to still be evolving, so it doesn't necessarily fit the goal of minimizing maintenance effort (which seems to be a driving factor here). All in all, I think it makes sense to support both Cluster API and this new provider and let the users choose.

Reply:
i agree. my point here was only to say that there are other options. i don't think it's fair to say that the only way for a user to add new providers is by forking the repo. there are other approaches, they are complex and have their own issues, but they do exist.

Reply:
> Perhaps most importantly Cluster API seems to still be evolving, so it doesn't necessarily fit the goal of minimizing maintenance effort (which seems to be a driving factor here). All in all, I think it makes sense to support both Cluster API and this new provider and let the users choose.

Yeah, I believe many people is still not able to transition to cluster-api and cluster-api is not mature enough yet.

@elmiko I'd mention that an alternative would be to use the cluster-api.

Reply:
> Yeah, I believe many people is still not able to transition to cluster-api and cluster-api is not mature enough yet.
> 

i agree, it is still working through alpha versions.

> @elmiko I'd mention that an alternative would be to use the cluster-api but I don't know if you have any example on how this could be possible.

i think that's all i'm asking for here, just changing the language so that it doesn't paint a black/white picture.  you might mention that cluster-api is another approach to this problem, i don't think it needs to be more than that.

Reply:
The integration with cluster-api could easily be made more generic.  The autoscaler could scale *something* and something else responds.  Today, that something is machineSets or machineDeployments, but it could easily be componentX.

Perhaps cluster-api implementation should be refactored to support a generic scalable resource.